### ISSUE TITLE
Unable to send message from method outside @ProtocolEntityCallback

### ISSUE THREAD

**Issue by shishigami** [tgalal/yowsup#717]
Hello, everybody. I'm making an application to receive messages from whatsapp and a message queue. The problem I'm experiencing is I cannot send messages from a method that's not decorated by __@ProtocolEntityCallback__. 

Here's a [Gist](https://gist.github.com/shishigami/b4222fa87858c5d2d335) with all the files composing the app.

What's the issue?
The code inside __send_message__ has been tested inside __@ProtocolEntityCallback("message")__ and worked fine. When the code has been added in both methods it runs twice, but when it's just inside __send_message__ it doesn't work at all.

What do I need to achieve?
I need to send messages that have been processed and stored inside __self._queue__.

Any help will be very much appreciated.
Thank you.

**Comment by shishigami** [comment]
Using logging inside **layer.py** and got this:

```xml
DEBUG:yowsup.stacks.yowstack:Initializing stack
DEBUG:yowsup.stacks.yowstack:Constructed Network Layer
DEBUG:yowsup.stacks.yowstack:Constructed Stanza Regulator Layer
DEBUG:yowsup.stacks.yowstack:Constructed Crypt Layer
DEBUG:yowsup.stacks.yowstack:Constructed Coder Layer
DEBUG:yowsup.stacks.yowstack:Constructed Logger Layer
DEBUG:yowsup.stacks.yowstack:Constructed Authentication Layer - Messages Layer - Receipt Layer - Ack Layer - Presence Layer - Ib Layer - Iq Layer - notification Ib Layer - Iq Layer - Chatstate Layer - Groups Iq Layer - Media Layer - Privacy Layer
COOL! HeadLayer started
DEBUG:yowsup.stacks.yowstack:Constructed Interface Layer
DEBUG:yowsup.layers.logger.layer:tx:
<stream:features>
<readreceipts>
</readreceipts>
<groups_v2>
</groups_v2>
<privacy>
</privacy>
<presence>
</presence>
</stream:features>

DEBUG:yowsup.layers.logger.layer:tx:
<auth passive="false" user="5492216397612" mechanism="WAUTH-2">
���{i0���p�-4��͍Ɵ ���T��A��R~o��Ow��M��X
HEX:94e4ab7b6930ec01a9b470cd2d3490f3cd8dc69f20ac83a55480af4101b79d527e6f99e44f77a2a18a084dfef258ed
</auth>

DEBUG:yowsup.layers.logger.layer:rx:
<stream:features>
</stream:features>

DEBUG:yowsup.layers.logger.layer:rx:
<success status="active" kind="free" creation="1426743244" expiration="1458279244" props="4" t="1429338747">
�G�R�g&5qjF��D�
HEX:9547ea521cc0126726358ee3711a6a46fcd24480
</success>

DEBUG:yowsup.layers.protocol_iq.layer:starting ping thread.
('Logged in!', 'Auth')
starting cerebro
COOL! WPEar started
COOL! MQEar started
DEBUG:yowsup.layers.logger.layer:rx:
<ib from="s.whatsapp.net">
<offline count="0">
</offline>
</ib>

there's a message: {'source': 'WP'}
there's a message: {'source': 'MQ'}
there's a message: {u'msg': u'this is a test and nothing more', u'phone': u'5492215899490@s.whatsapp.net'}
DEBUG:yowsup.layers.logger.layer:tx:
<message to="5492215899490@s.whatsapp.net" type="text" id="1429338559-1">
<body>
THIS IS A TEST AND NOTHING MORE
HEX:544849532049532041205445535420414e44204e4f5448494e47204d4f5245
</body>
</message>

DEBUG:yowsup.layers.logger.layer:rx:
<ack t="1429338747" from="5492215899490@s.whatsapp.net" id="1429338559-1" class="message">
</ack>

DEBUG:yowsup.layers.logger.layer:rx:
<receipt from="5492215899490@s.whatsapp.net" id="1429338559-1" t="1429338747">
</receipt>

DEBUG:yowsup.layers.logger.layer:tx:
<ack type="delivery" id="1429338559-1" class="receipt">
</ack>

DEBUG:yowsup.layers.logger.layer:rx:
<stream:error>
<xml-not-well-formed>
</xml-not-well-formed>
</stream:error>

DEBUG:yowsup.layers.network.layer:Disconnected, reason: Connection Closed
DEBUG:yowsup.layers.network.layer:Disconnected, reason: Connection Closed
```

The message was successfully sent, but the app shuts automatically shuts down after the ```<xml-not-well-formed>```.

The [Gist](https://gist.github.com/shishigami/b4222fa87858c5d2d335) has been updated with the the current changes made to **layer.py**.

The library has become easier to understand after digging through its files and it's an awesome piece of work. But there are some things (like this problem I'm having) that turn hard to solve. I'm looking forward to solve it by myself, but if there's a tip to guide me a step closer to the solution will be much appreciated.

**Comment by shishigami** [comment]
the problem was solved by using __threading__ instead of __multiprocessing__ and blocking message send until all acks were received.

**Comment by ComFreek** [comment]
@shishigami I am currently facing the same problems you had. I would really like to continue using multiprocessing because threading does only provide IO concurrency and not real CPU usage concurrency. Had you been able to truly hold multiprocessing liable to cause the issue or was your success rather a side effect of waiting for the receipt of all acks before sending?

How did you end up implementing the latter? To be honest, I don't understand which acks you are referring to. As far as I know only clients are supposed to send them to the server indicating the acknowledgement of a receipt (see [Yowsup 2.0 Sample app/layer.py IV](https://github.com/tgalal/yowsup/wiki/Yowsup-2.0-Sample-app#layerpy-iv) in the wiki). Then, how come that you have to wait for them &mdash; from the point of view of a client?

**Comment by shishigami** [comment]
the code version is former to may 1st, when whatsapp changed the ack params to include the receiver's whatsapp ID, that's why the ack I send in "receipt" callback is different from the yowsup demo sample.

I haven't been able to use multiprocessing, I had to switch to threading. after that change, everything worked. also had to get rid of adjacent modules I have created. so I had to settle with IO concurrency instead of CPU.

sorry if I can't give more help
